File: snobot_sim/cpp/lib/SnobotSim/SensorActuatorRegistry.cpp
/*
 * SensorActuatorRegistry.cpp
 *
 *  Created on: May 3, 2017
 *      Author: PJ
 */

#include "SnobotSim/SensorActuatorRegistry.h"
#include "SnobotSim/SimulatorComponents/CompressorWrapper.h"

SensorActuatorRegistry* SensorActuatorRegistry::sInstance =
        new SensorActuatorRegistry();


SensorActuatorRegistry& SensorActuatorRegistry::Get()
{
	return *sInstance;
}
    
SensorActuatorRegistry::SensorActuatorRegistry() :
        mCompressor(new CompressorWrapper)
{

}

SensorActuatorRegistry::~SensorActuatorRegistry()
{
    Reset();
}


void SensorActuatorRegistry::Reset()
{
    mSpeedControllerWrapperMap.clear();
    mRelayWrapperMap.clear();
    mDigitalSourceWrapperMap.clear();
    mAnalogSourceWrapperMap.clear();
    mSolenoidWrapperMap.clear();
    mEncoderWrapperMap.clear();
    mGyroWrapperMap.clear();
    mAccelerometerWrapperMap.clear();
    mISpiWrapperMap.clear();
    mII2CWrapperMap.clear();

    mSimulatorComponents.clear();
}

std::shared_ptr<CompressorWrapper> SensorActuatorRegistry::GetCompressorWrapper()
{
    return mCompressor;
}

void SensorActuatorRegistry::AddSimulatorComponent(const std::shared_ptr<ISimulatorUpdater>& aSimulatorComponent)
{
    mSimulatorComponents.push_back(aSimulatorComponent);
}


std::vector<std::shared_ptr<ISimulatorUpdater>>& SensorActuatorRegistry::GetSimulatorComponents()
{
    return mSimulatorComponents;
}


#define ACTUATOR_GETTERS(ItemType)                                                                              \
    bool SensorActuatorRegistry::Register(int aPort,  const std::shared_ptr<ItemType>& aActuator)               \
    {                                                                                                           \
        return RegisterItem(aPort, aActuator, m##ItemType##Map, #ItemType);                                     \
    }                                                                                                           \
    std::shared_ptr<ItemType> SensorActuatorRegistry::Get##ItemType(int aPort, bool aLogError)                  \
    {                                                                                                           \
        return GetItem(aPort, m##ItemType##Map, #ItemType, aLogError);                                          \
    }                                                                                                           \
    const std::map<int, std::shared_ptr<ItemType>>& SensorActuatorRegistry::Get##ItemType##Map() const          \
    {                                                                                                           \
        return m##ItemType##Map;                                                                                \
    }                                                                                                           \
    std::map<int, std::shared_ptr<ItemType>>& SensorActuatorRegistry::Get##ItemType##Map()                      \
    {                                                                                                           \
        return m##ItemType##Map;                                                                                \
    }                                                                                                           \

ACTUATOR_GETTERS(SpeedControllerWrapper)
ACTUATOR_GETTERS(RelayWrapper)
ACTUATOR_GETTERS(DigitalSourceWrapper)
ACTUATOR_GETTERS(AnalogSourceWrapper)
ACTUATOR_GETTERS(SolenoidWrapper)
ACTUATOR_GETTERS(EncoderWrapper)
ACTUATOR_GETTERS(AccelerometerWrapper)
ACTUATOR_GETTERS(GyroWrapper)
ACTUATOR_GETTERS(ISpiWrapper)
ACTUATOR_GETTERS(II2CWrapper)

File: snobot_sim_java/java/snobot_sim_java/com/snobot/simulator/SensorActuatorRegistry.java
package com.snobot.simulator;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import com.snobot.simulator.module_wrapper.AnalogWrapper;
import com.snobot.simulator.module_wrapper.DigitalSourceWrapper;
import com.snobot.simulator.module_wrapper.EncoderWrapper;
import com.snobot.simulator.module_wrapper.PwmWrapper;
import com.snobot.simulator.module_wrapper.RelayWrapper;
import com.snobot.simulator.module_wrapper.SolenoidWrapper;
import com.snobot.simulator.simulator_components.II2CWrapper;
import com.snobot.simulator.simulator_components.ISimulatorUpdater;
import com.snobot.simulator.simulator_components.ISpiWrapper;
import com.snobot.simulator.simulator_components.accelerometer.AccelerometerWrapper;
import com.snobot.simulator.simulator_components.gyro.GyroWrapper;

public class SensorActuatorRegistry
{
    private static SensorActuatorRegistry mInstance = new SensorActuatorRegistry();

    private Map<Integer, PwmWrapper> mSpeedControllerMap = new HashMap<>();
    private Map<Integer, RelayWrapper> mRelayWrapperMap = new HashMap<>();
    private Map<Integer, DigitalSourceWrapper> mDigitalSourceWrapperMap = new HashMap<>();
    private Map<Integer, AnalogWrapper> mAnalogSourceWrapperMap = new HashMap<>();
    private Map<Integer, SolenoidWrapper> mSolenoidWrapperMap = new HashMap<>();
    private Map<Integer, EncoderWrapper> mEncoderWrapperMap = new HashMap<>();

    private Map<Integer, GyroWrapper> mGyroWrapperMap = new HashMap<>();
    private Map<Integer, AccelerometerWrapper> mAccelerometerWrapperMap = new HashMap<>();
    private Map<Integer, II2CWrapper> mI2CWrapperMap = new HashMap<>();
    private Map<Integer, ISpiWrapper> mSpiWrapperMap = new HashMap<>();

    private Collection<ISimulatorUpdater> mSimulatorComponents = new ArrayList<>();

    private SensorActuatorRegistry()
    {

    }

    public static SensorActuatorRegistry get()
    {
        return mInstance;
    }

    public <ItemType> boolean registerItem(ItemType aItem, int aPort, Map<Integer, ItemType> aMap, String aMessage)
    {
        aMap.put(aPort, aItem);
        return true;
    }

    public boolean register(AnalogWrapper aActuator, int aPort)
    {
        return registerItem(aActuator, aPort, mAnalogSourceWrapperMap, "Analog");
    }

    public boolean register(PwmWrapper aActuator, int aPort)
    {
        return registerItem(aActuator, aPort, mSpeedControllerMap, "Speed Controller");
    }

    public boolean register(DigitalSourceWrapper aSensor, int aPort)
    {
        return registerItem(aSensor, aPort, mDigitalSourceWrapperMap, "Digital IO");
    }

    public boolean register(SolenoidWrapper aActuator, int aPort)
    {
        return registerItem(aActuator, aPort, mSolenoidWrapperMap, "Solenoid");
    }

    public boolean register(RelayWrapper aActuator, int aPort)
    {
        return registerItem(aActuator, aPort, mRelayWrapperMap, "Relay");
    }

    public boolean register(EncoderWrapper aEncoder, Integer aPort)
    {
        return registerItem(aEncoder, aPort, mEncoderWrapperMap, "Encoder");
    }

    public boolean register(AccelerometerWrapper aSensor, Integer aPort)
    {
        return registerItem(aSensor, aPort, mAccelerometerWrapperMap, "Accelerometer");
    }

    public boolean register(GyroWrapper aSensor, Integer aPort)
    {
        return registerItem(aSensor, aPort, mGyroWrapperMap, "Gyro");
    }

    public boolean register(II2CWrapper aSensor, Integer aPort)
    {
        return registerItem(aSensor, aPort, mI2CWrapperMap, "I2C");
    }

    public boolean register(ISpiWrapper aSensor, Integer aPort)
    {
        return registerItem(aSensor, aPort, mSpiWrapperMap, "SPI");
    }

    public boolean register(ISimulatorUpdater aUpdater)
    {
        mSimulatorComponents.add(aUpdater);
        return true;
    }

    public Map<Integer, PwmWrapper> getSpeedControllers()
    {
        return mSpeedControllerMap;
    }

    public Map<Integer, SolenoidWrapper> getSolenoids()
    {
        return mSolenoidWrapperMap;
    }

    public Map<Integer, DigitalSourceWrapper> getDigitalSources()
    {
        return mDigitalSourceWrapperMap;
    }

    public Map<Integer, RelayWrapper> getRelays()
    {
        return mRelayWrapperMap;
    }

    public Map<Integer, AnalogWrapper> getAnalog()
    {
        return mAnalogSourceWrapperMap;
    }

    public Map<Integer, EncoderWrapper> getEncoders()
    {
        return mEncoderWrapperMap;
    }

    public Map<Integer, AccelerometerWrapper> getAccelerometers()
    {
        return mAccelerometerWrapperMap;
    }

    public Map<Integer, GyroWrapper> getGyros()
    {
        return mGyroWrapperMap;
    }

    public Map<Integer, II2CWrapper> getI2CWrapperss()
    {
        return mI2CWrapperMap;
    }

    public Map<Integer, ISpiWrapper> getSpiWrappers()
    {
        return mSpiWrapperMap;
    }

    public Collection<ISimulatorUpdater> getSimulatorComponents()
    {
        return mSimulatorComponents;
    }

    public void reset()
    {
        mSpeedControllerMap.clear();
        mRelayWrapperMap.clear();
        mDigitalSourceWrapperMap.clear();
        mAnalogSourceWrapperMap.clear();
        mSolenoidWrapperMap.clear();
        mEncoderWrapperMap.clear();

        mGyroWrapperMap.clear();
        mAccelerometerWrapperMap.clear();
        mI2CWrapperMap.clear();
        mSpiWrapperMap.clear();

        mSimulatorComponents.clear();
    }
}
